phpBB/phpbb/storage/controller/attachment.php

```diff
@@ -0,0 +1,494 @@
+			}
+		}
+
+		$redirect = '';
+
+		/**
+		 * Event to modify data before sending file to browser
+		 *
+		 * @event core.download_file_send_to_browser_before
+		 * @var	int		attach_id			The attachment ID
+		 * @var	array	attachment			Array with attachment data
+		 * @var	array	extensions			Array with file extensions data
+		 * @var	bool	thumbnail			Flag indicating if the file is a thumbnail
+		 * @var	string	redirect			Do a redirection instead of reading the file
+		 * @since 3.1.6-RC1
+		 * @changed 3.1.7-RC1	Fixing wrong name of a variable (replacing "extension" by "extensions")
+		 * @changed 3.3.0-a1		Add redirect variable
+		 * @changed 3.3.0-a1		Remove display_cat variable
+		 * @changed 3.3.0-a1		Remove mode variable
+		 */
+		$vars = array(
+			'attach_id',
+			'attachment',
+			'extensions',
+			'thumbnail',
+			'redirect',
+		);
+		extract($this->dispatcher->trigger_event('core.download_file_send_to_browser_before', compact($vars)));
+
+		// If the redirect variable have been overwritten, do redirect there
+		if (!empty($redirect))
+		{
+			return new RedirectResponse($redirect);
+		}
+
+		// Check if the file exists in the storage table too
+		if (!$this->storage->exists($attachment['physical_filename']))
+		{
+			throw new http_exception(404, 'ERROR_NO_ATTACHMENT');
+		}
+
+		/**
+		 * Event to alter attachment before it is sent to browser.
+		 *
+		 * @event core.send_file_to_browser_before
+		 * @var	array	attachment	Attachment data
+		 * @since 3.1.11-RC1
+		 * @changed 3.3.0-a1		Removed category variable
+		 * @changed 3.3.0-a1		Removed size variable
+		 * @changed 3.3.0-a1		Removed filename variable
+		 */
+		$vars = array(
+			'attachment',
+		);
+		extract($this->dispatcher->trigger_event('core.send_file_to_browser_before', compact($vars)));
+
+		// Content-type header
+		$this->response->headers->set('Content-Type', $attachment['mimetype']);
+
+		// Display images in browser and force download for other file types
+		if (strpos($attachment['mimetype'], 'image') !== false)
+		{
+			$disposition = $this->response->headers->makeDisposition(
+				ResponseHeaderBag::DISPOSITION_INLINE,
+				rawurlencode(htmlspecialchars_decode($attachment['real_filename']))
+			);
+		}
+		else
+		{
+			$disposition = $this->response->headers->makeDisposition(
+				ResponseHeaderBag::DISPOSITION_ATTACHMENT,
+				rawurlencode(htmlspecialchars_decode($attachment['real_filename']))
+			);
+		}
+
+		$this->response->headers->set('Content-Disposition', $disposition);
+
+		// Set expires header for browser cache
+		$time = new \Datetime();
+		$this->response->setExpires($time->modify('+1 year'));
+
+		return parent::handle($attachment['physical_filename']);
+	}
+
+	/**
+	 * Handles authentication when downloading attachments from a post or topic
+	 *
+	 * @param int $topic_id The id of the topic that we are downloading from
+	 *
+	 * @return null
+	 */
+	protected function phpbb_download_handle_forum_auth($topic_id)
```

Comment:
There is really no need to use such a crappy name anymore since it's part of the controller class ;)

Reply:
This is a copy/paste from the functions to handle attachment download permissions. Nicofuma said that this should be moved to other place in the future